clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedInternalCommandConsumer.java

```diff
@@ -0,0 +1,191 @@
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.client.command.amqp;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.qpid.proton.amqp.messaging.Rejected;
+import org.apache.qpid.proton.amqp.transport.ErrorCondition;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.adapter.client.amqp.AbstractServiceClient;
+import org.eclipse.hono.adapter.client.command.CommandContext;
+import org.eclipse.hono.adapter.client.command.CommandHandlerWrapper;
+import org.eclipse.hono.adapter.client.command.CommandHandlers;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.SendMessageSampler;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.Lifecycle;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.vertx.core.Future;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+
+/**
+ * A vertx-proton based consumer to receive commands forwarded by the Command Router on the internal command endpoint.
+ */
+public class ProtonBasedInternalCommandConsumer extends AbstractServiceClient implements Lifecycle {
+
+    protected static final Logger LOG = LoggerFactory.getLogger(ProtonBasedInternalCommandConsumer.class);
+
+    private static final int RECREATE_CONSUMER_DELAY = 20;
+
+    private final String adapterInstanceId;
+    private final CommandHandlers commandHandlers;
+    private final AtomicBoolean recreatingConsumer = new AtomicBoolean(false);
+    private final AtomicBoolean tryAgainRecreatingConsumer = new AtomicBoolean(false);
+    private final Tracer tracer;
+
+    private ProtonReceiver adapterSpecificConsumer;
+
+    /**
+     * Creates a new consumer for an existing connection.
+     *
+     * @param connection The connection to the AMQP network.
+     * @param adapterInstanceId The adapter instance id.
+     * @param commandHandlers The command handlers to choose from for handling a received command.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public ProtonBasedInternalCommandConsumer(
+            final HonoConnection connection,
+            final String adapterInstanceId,
+            final CommandHandlers commandHandlers) {
+        super(connection, SendMessageSampler.Factory.noop());
+
+        this.commandHandlers = Objects.requireNonNull(commandHandlers);
+        this.adapterInstanceId = Objects.requireNonNull(adapterInstanceId);
+        tracer = connection.getTracer();
+    }
+
+    @Override
+    public Future<Void> start() {
+        return super.start()
```

Comment:
FMPOV we should not rely on any particular behavior of the super class' method's default implementation here. The *start* method is explicitly designed to be overridden and its default implementation could therefore be changed at any time as well. IMHO we should therefore make sure to start the connection here ourselves ...

Reply:
I think in the particular case of the `AbstractServiceClient` class, all its sub classes rely on the fact that invocation of `connection.connect()` and `connection.disconnect()` on startup and shutdown is done in `AbstractServiceClient`, so that there is no need to invoke these methods in start()/stop() in the sub classes.
In that sense, I would rather make this behaviour more explicit in the javadoc of `AbstractServiceClient` and then keep the invocation of `super.start()` in `ProtonBasedInternalCommandConsumer`.

Otherwise, if I explicitly call `connection.connect()` in `ProtonBasedInternalCommandConsumer#start`, it would look odd not to override the `stop()` method as well, calling `connection.disconnect()` there. But this method would then be the same as in `AbstractServiceClient`.

Reply:
I've adapted the javadoc of the start/stop methods of `AbstractServiceClient` to make the connect/disconnect invocation there more explicit.

Reply:
I still do not agree with invoking super.start() here. Overriding AbstractServiceClient.start() without overriding AbstractServiceClient.stop() is perfectly alright IMHO as it simply means that the implementation of AbtstractServiceClient.stop() is still in effect and behaves as specified (in AbstractServiceClient).

But I will not be religious about it ...